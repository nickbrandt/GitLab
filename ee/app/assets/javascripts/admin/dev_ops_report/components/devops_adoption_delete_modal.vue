<script>
import { GlModal, GlSprintf, GlAlert } from '@gitlab/ui';
import * as Sentry from '~/sentry/wrapper';
import { DEVOPS_ADOPTION_STRINGS, DEVOPS_ADOPTION_SEGMENT_DELETE_MODAL_ID } from '../constants';
import deleteDevopsAdoptionSegmentMutation from '../graphql/mutations/delete_devops_adoption_segment.mutation.graphql';
import { deleteSegmentFromCache } from '../utils/cache_updates';

export default {
  name: 'DevopsAdoptionDeleteModal',
  components: { GlModal, GlSprintf, GlAlert },
  i18n: DEVOPS_ADOPTION_STRINGS.deleteModal,
  devopsSegmentDeleteModalId: DEVOPS_ADOPTION_SEGMENT_DELETE_MODAL_ID,
  props: {
    segment: {
      type: Object,
      required: true,
    },
  },
  data() {
    return {
      loading: false,
      errors: [],
    };
  },
  computed: {
    cancelOptions() {
      return {
        text: this.$options.i18n.cancel,
        attributes: [{ disabled: this.loading }],
      };
    },
    primaryOptions() {
      return {
        text: this.$options.i18n.confirm,
        attributes: [
          {
            variant: 'danger',
            loading: this.loading,
          },
        ],
      };
    },
    displayError() {
      return this.errors[0];
    },
    displayName() {
      return this.segment.namespace?.fullName;
    },
  },
  methods: {
    async deleteSegment() {
      try {
        const {
          segment: { id },
        } = this;

        this.loading = true;

        const {
          data: {
            deleteDevopsAdoptionSegment: { errors },
          },
        } = await this.$apollo.mutate({
          mutation: deleteDevopsAdoptionSegmentMutation,
          variables: {
            id,
          },
          update(store) {
            deleteSegmentFromCache(store, id);
          },
        });

        if (errors.length) {
          this.errors = errors;
        } else {
          this.$refs.modal.hide();
        }
      } catch (error) {
        this.errors.push(this.$options.i18n.error);
        Sentry.captureException(error);
      } finally {
        this.loading = false;
      }
    },
    clearErrors() {
      this.errors = [];
    },
  },
};
</script>
<template>
  <gl-modal
    ref="modal"
    :modal-id="$options.devopsSegmentDeleteModalId"
    size="sm"
    :action-primary="primaryOptions"
    :action-cancel="cancelOptions"
    @primary.prevent="deleteSegment"
  >
    <template #modal-title>{{ $options.i18n.title }}</template>
    <gl-alert v-if="errors.length" variant="danger" class="gl-mb-3" @dismiss="clearErrors">
      {{ displayError }}
    </gl-alert>
    <gl-sprintf :message="$options.i18n.confirmationMessage">
      <template #name
        ><strong>{{ displayName }}</strong></template
      >
    </gl-sprintf>
  </gl-modal>
</template>
