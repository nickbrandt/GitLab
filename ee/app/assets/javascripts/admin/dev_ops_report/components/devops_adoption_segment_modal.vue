<script>
import { GlFormGroup, GlFormInput, GlFormRadioGroup, GlModal, GlAlert, GlIcon } from '@gitlab/ui';
import { convertToGraphQLId, TYPE_GROUP } from '~/graphql_shared/utils';
import * as Sentry from '~/sentry/wrapper';
import createDevopsAdoptionSegmentMutation from '../graphql/mutations/create_devops_adoption_segment.mutation.graphql';
import { DEVOPS_ADOPTION_STRINGS, DEVOPS_ADOPTION_SEGMENT_MODAL_ID } from '../constants';
import { addSegmentToCache } from '../utils/cache_updates';

export default {
  name: 'DevopsAdoptionSegmentModal',
  components: {
    GlModal,
    GlFormGroup,
    GlFormInput,
    GlFormRadioGroup,
    GlAlert,
    GlIcon,
  },
  props: {
    groups: {
      type: Array,
      required: true,
    },
  },
  i18n: DEVOPS_ADOPTION_STRINGS.modal,
  data() {
    return {
      selectedGroupId: null,
      filter: '',
      loading: false,
      errors: [],
    };
  },
  computed: {
    checkboxOptions() {
      return this.groups.map(({ id, full_name }) => ({ text: full_name, value: id }));
    },
    cancelOptions() {
      return {
        button: {
          text: this.$options.i18n.cancel,
          attributes: [{ disabled: this.loading }],
        },
        callback: this.resetForm,
      };
    },
    primaryOptions() {
      return {
        button: {
          text: this.$options.i18n.addingButton,
          attributes: [
            {
              variant: 'info',
              loading: this.loading,
              disabled: !this.canSubmit,
            },
          ],
        },
        callback: this.createSegment,
      };
    },
    canSubmit() {
      return Boolean(this.selectedGroupId);
    },
    displayError() {
      return this.errors[0];
    },
    modalTitle() {
      return this.$options.i18n.addingTitle;
    },
    filteredOptions() {
      return this.filter
        ? this.checkboxOptions.filter((option) =>
            option.text.toLowerCase().includes(this.filter.toLowerCase()),
          )
        : this.checkboxOptions;
    },
  },
  methods: {
    async createSegment() {
      try {
        this.loading = true;
        const {
          data: {
            createDevopsAdoptionSegment: { errors },
          },
        } = await this.$apollo.mutate({
          mutation: createDevopsAdoptionSegmentMutation,
          variables: {
            namespaceId: convertToGraphQLId(TYPE_GROUP, this.selectedGroupId),
          },
          update: (store, { data }) => {
            const {
              createDevopsAdoptionSegment: { segment, errors: requestErrors },
            } = data;

            if (!requestErrors.length) addSegmentToCache(store, segment);
          },
        });

        if (errors.length) {
          this.errors = errors;
        } else {
          this.resetForm();
          this.closeModal();
        }
      } catch (error) {
        this.errors.push(this.$options.i18n.error);
        Sentry.captureException(error);
      } finally {
        this.loading = false;
      }
    },
    clearErrors() {
      this.errors = [];
    },
    closeModal() {
      this.$refs.modal.hide();
    },
    resetForm() {
      this.selectedGroupId = null;
      this.filter = '';
    },
  },
  devopsSegmentModalId: DEVOPS_ADOPTION_SEGMENT_MODAL_ID,
};
</script>
<template>
  <gl-modal
    ref="modal"
    :modal-id="$options.devopsSegmentModalId"
    :title="modalTitle"
    size="sm"
    scrollable
    :action-primary="primaryOptions.button"
    :action-cancel="cancelOptions.button"
    @primary.prevent="primaryOptions.callback"
    @canceled="cancelOptions.callback"
    @hide="resetForm"
  >
    <gl-alert v-if="errors.length" variant="danger" class="gl-mb-3" @dismiss="clearErrors">
      {{ displayError }}
    </gl-alert>
    <gl-form-group class="gl-mb-3" data-testid="filter">
      <gl-icon name="search" :size="18" class="gl-text-gray-300 gl-absolute gl-mt-3 gl-ml-3" />
      <gl-form-input
        v-model="filter"
        class="gl-pl-7!"
        type="text"
        :placeholder="$options.i18n.filterPlaceholder"
        :disabled="loading"
      />
    </gl-form-group>
    <gl-form-group class="gl-mb-0">
      <gl-form-radio-group
        v-if="filteredOptions.length"
        :key="filteredOptions.length"
        v-model="selectedGroupId"
        data-testid="groups"
        :options="filteredOptions"
        :hide-toggle-all="true"
        :disabled="loading"
        class="gl-p-3 gl-pb-0 gl-mb-2 gl-border-1 gl-border-solid gl-border-gray-100 gl-rounded-base"
      />
      <gl-alert v-else variant="info" :dismissible="false" data-testid="filter-warning">
        {{ $options.i18n.noResults }}
      </gl-alert>
    </gl-form-group>
  </gl-modal>
</template>
