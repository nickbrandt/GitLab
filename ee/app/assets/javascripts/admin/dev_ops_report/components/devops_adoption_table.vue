<script>
import {
  GlTable,
  GlButton,
  GlPopover,
  GlModalDirective,
  GlTooltipDirective,
  GlIcon,
} from '@gitlab/ui';
import LocalStorageSync from '~/vue_shared/components/local_storage_sync.vue';
import {
  DEVOPS_ADOPTION_TABLE_TEST_IDS,
  DEVOPS_ADOPTION_STRINGS,
  DEVOPS_ADOPTION_SEGMENT_MODAL_ID,
  DEVOPS_ADOPTION_SEGMENT_DELETE_MODAL_ID,
  DEVOPS_ADOPTION_SEGMENTS_TABLE_SORT_BY_STORAGE_KEY,
  DEVOPS_ADOPTION_SEGMENTS_TABLE_SORT_DESC_STORAGE_KEY,
} from '../constants';
import DevopsAdoptionTableCellFlag from './devops_adoption_table_cell_flag.vue';
import DevopsAdoptionDeleteModal from './devops_adoption_delete_modal.vue';

const NAME_HEADER = 'name';

const formatter = (value, key, item) => {
  if (key === NAME_HEADER) {
    return item.namespace?.fullName;
  }

  if (item.latestSnapshot && item.latestSnapshot[key] === false) {
    return 1;
  } else if (item.latestSnapshot && item.latestSnapshot[key]) {
    return 2;
  }

  return 0;
};

const fieldOptions = {
  thClass: 'gl-bg-white! gl-text-gray-400',
  thAttr: { 'data-testid': DEVOPS_ADOPTION_TABLE_TEST_IDS.TABLE_HEADERS },
  formatter,
  sortable: true,
  sortByFormatted: true,
};

const { table: i18n } = DEVOPS_ADOPTION_STRINGS;

const headers = [
  NAME_HEADER,
  'issueOpened',
  'mergeRequestOpened',
  'mergeRequestApproved',
  'runnerConfigured',
  'pipelineSucceeded',
  'deploySucceeded',
  'securityScanSucceeded',
].map((key) => ({ key, ...i18n.headers[key], ...fieldOptions }));

export default {
  name: 'DevopsAdoptionTable',
  components: {
    GlTable,
    DevopsAdoptionTableCellFlag,
    GlButton,
    GlPopover,
    LocalStorageSync,
    DevopsAdoptionDeleteModal,
    GlIcon,
  },
  i18n,
  devopsSegmentModalId: DEVOPS_ADOPTION_SEGMENT_MODAL_ID,
  devopsSegmentDeleteModalId: DEVOPS_ADOPTION_SEGMENT_DELETE_MODAL_ID,
  directives: {
    GlTooltip: GlTooltipDirective,
    GlModal: GlModalDirective,
  },
  tableHeaderFields: [
    ...headers,
    {
      key: 'actions',
      tdClass: 'actions-cell',
      ...fieldOptions,
      sortable: false,
    },
  ],
  testids: DEVOPS_ADOPTION_TABLE_TEST_IDS,
  sortByStorageKey: DEVOPS_ADOPTION_SEGMENTS_TABLE_SORT_BY_STORAGE_KEY,
  sortDescStorageKey: DEVOPS_ADOPTION_SEGMENTS_TABLE_SORT_DESC_STORAGE_KEY,
  props: {
    segments: {
      type: Array,
      required: true,
    },
    selectedSegment: {
      type: Object,
      required: false,
      default: null,
    },
  },
  data() {
    return {
      sortBy: NAME_HEADER,
      sortDesc: false,
    };
  },
  methods: {
    popoverContainerId(name) {
      return `popover_container_id_for_${name}`;
    },
    popoverId(name) {
      return `popover_id_for_${name}`;
    },
    setSelectedSegment(segment) {
      this.$emit('set-selected-segment', segment);
    },
    slotName(key) {
      return `head(${key})`;
    },
  },
};
</script>
<template>
  <div>
    <local-storage-sync
      v-model="sortBy"
      :storage-key="$options.sortByStorageKey"
      :data-testid="$options.testids.LOCAL_STORAGE_SORT_BY"
      as-json
    />
    <local-storage-sync
      v-model="sortDesc"
      :storage-key="$options.sortDescStorageKey"
      :data-testid="$options.testids.LOCAL_STORAGE_SORT_DESC"
      as-json
    />
    <gl-table
      :fields="$options.tableHeaderFields"
      :items="segments"
      :sort-by.sync="sortBy"
      :sort-desc.sync="sortDesc"
      thead-class="gl-border-t-0 gl-border-b-solid gl-border-b-1 gl-border-b-gray-100"
      stacked="sm"
    >
      <template v-for="header in $options.tableHeaderFields" #[slotName(header.key)]>
        <div :key="header.key" class="gl-display-flex gl-align-items-center">
          <span>{{ header.label }}</span>
          <gl-icon
            v-if="header.tooltip"
            v-gl-tooltip.hover="header.tooltip"
            name="information-o"
            class="gl-text-gray-200 gl-ml-1"
            :size="14"
          />
        </div>
      </template>

      <template
        #cell(name)="{
          item: {
            namespace: { fullName },
            latestSnapshot,
          },
        }"
      >
        <div :data-testid="$options.testids.SEGMENT">
          <strong v-if="latestSnapshot">{{ fullName }}</strong>
          <template v-else>
            <span class="gl-text-gray-400">{{ fullName }}</span>
            <gl-icon name="hourglass" class="gl-text-gray-400" />
          </template>
        </div>
      </template>

      <template #cell(issueOpened)="{ item }">
        <devops-adoption-table-cell-flag
          v-if="item.latestSnapshot"
          :data-testid="$options.testids.ISSUES"
          :enabled="item.latestSnapshot.issueOpened"
        />
      </template>

      <template #cell(mergeRequestOpened)="{ item }">
        <devops-adoption-table-cell-flag
          v-if="item.latestSnapshot"
          :data-testid="$options.testids.MRS"
          :enabled="item.latestSnapshot.mergeRequestOpened"
        />
      </template>

      <template #cell(mergeRequestApproved)="{ item }">
        <devops-adoption-table-cell-flag
          v-if="item.latestSnapshot"
          :data-testid="$options.testids.APPROVALS"
          :enabled="item.latestSnapshot.mergeRequestApproved"
        />
      </template>

      <template #cell(runnerConfigured)="{ item }">
        <devops-adoption-table-cell-flag
          v-if="item.latestSnapshot"
          :data-testid="$options.testids.RUNNERS"
          :enabled="item.latestSnapshot.runnerConfigured"
        />
      </template>

      <template #cell(pipelineSucceeded)="{ item }">
        <devops-adoption-table-cell-flag
          v-if="item.latestSnapshot"
          :data-testid="$options.testids.PIPELINES"
          :enabled="item.latestSnapshot.pipelineSucceeded"
        />
      </template>

      <template #cell(deploySucceeded)="{ item }">
        <devops-adoption-table-cell-flag
          v-if="item.latestSnapshot"
          :data-testid="$options.testids.DEPLOYS"
          :enabled="item.latestSnapshot.deploySucceeded"
        />
      </template>

      <template #cell(securityScanSucceeded)="{ item }">
        <devops-adoption-table-cell-flag
          v-if="item.latestSnapshot"
          :data-testid="$options.testids.SCANNING"
          :enabled="item.latestSnapshot.securityScanSucceeded"
        />
      </template>

      <template #cell(actions)="{ item }">
        <div :data-testid="$options.testids.ACTIONS">
          <gl-button :id="popoverId(item.id)" category="tertiary" icon="ellipsis_h" />
          <div :id="popoverContainerId(item.id)">
            <gl-popover
              :target="popoverId(item.id)"
              :container="popoverContainerId(item.id)"
              triggers="hover focus"
              placement="left"
            >
              <gl-button
                v-gl-modal="$options.devopsSegmentDeleteModalId"
                category="tertiary"
                variant="danger"
                @click="setSelectedSegment(item)"
                >{{ $options.i18n.deleteButton }}</gl-button
              >
            </gl-popover>
          </div>
        </div>
      </template>
    </gl-table>
    <devops-adoption-delete-modal v-if="selectedSegment" :segment="selectedSegment" />
  </div>
</template>
