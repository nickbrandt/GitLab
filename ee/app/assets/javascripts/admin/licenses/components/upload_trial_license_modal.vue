<script>
import { GlModal, GlLink, GlIcon, GlSafeHtmlDirective as SafeHtml } from '@gitlab/ui';
import trialLicenseActivatedSvg from 'ee_icons/_ee_trial_license_activated.svg';
import csrf from '~/lib/utils/csrf';
import { __ } from '~/locale';

export default {
  name: 'UploadTrialLicenseModal',
  components: {
    GlModal,
    GlLink,
    GlIcon,
  },
  directives: {
    SafeHtml,
  },
  props: {
    licenseKey: {
      type: String,
      required: true,
    },
    adminLicensePath: {
      type: String,
      required: true,
    },
    initialShow: {
      type: Boolean,
      required: false,
      default: true,
    },
  },
  data() {
    return {
      visible: this.initialShow,
    };
  },
  methods: {
    submitForm() {
      this.$refs.form.submit();
    },
  },
  csrf,
  cancelOptions: {
    text: __('Cancel'),
  },
  primaryOptions: {
    text: __('Install license'),
    attributes: [
      {
        variant: 'confirm',
        category: 'primary',
      },
    ],
  },
  trialLicenseActivatedSvg,
};
</script>

<template>
  <gl-modal
    ref="modal"
    :visible="visible"
    modal-id="modal-upload-trial-license"
    body-class="modal-upload-trial-license"
    :action-primary="$options.primaryOptions"
    :action-cancel="$options.cancelOptions"
    @primary.prevent="submitForm"
  >
    <div class="trial-activated-graphic gl-display-flex gl-justify-content-center gl-mt-5">
      <span v-safe-html="$options.trialLicenseActivatedSvg" class="svg-container"></span>
    </div>
    <h3 class="gl-text-center">{{ __('Your trial license was issued!') }}</h3>
    <p class="gl-text-center gl-text-gray-500 mw-70p m-auto gl-font-size-h2 gl-line-height-28">
      {{
        __(
          'Your trial license was issued and activated. Install it to enjoy GitLab Ultimate for 30 days.',
        )
      }}
    </p>
    <form
      id="new_license"
      ref="form"
      :action="adminLicensePath"
      enctype="multipart/form-data"
      method="post"
    >
      <div class="gl-mt-5">
        <gl-link
          id="hide-license"
          href="#hide-license"
          class="hide-license gl-text-gray-600 gl-text-center"
          >{{ __('Show license key') }}<gl-icon name="chevron-down"
        /></gl-link>
        <gl-link
          id="show-license"
          href="#show-license"
          class="show-license gl-text-gray-600 gl-text-center"
          >{{ __('Hide license key') }}<gl-icon name="chevron-up"
        /></gl-link>
        <div class="card trial-license-preview gl-overflow-y-auto gl-word-break-all gl-mt-5">
          {{ licenseKey }}
        </div>
      </div>
      <input :value="$options.csrf.token" type="hidden" name="authenticity_token" />
      <input id="license_data" :value="licenseKey" type="hidden" name="license[data]" />
    </form>
  </gl-modal>
</template>
