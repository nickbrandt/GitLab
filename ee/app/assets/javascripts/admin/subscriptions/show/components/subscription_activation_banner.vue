<script>
import { GlBanner, GlLink, GlSprintf } from '@gitlab/ui';
import {
  activateCloudLicense,
  subscriptionBannerText,
  subscriptionBannerTitle,
} from '../constants';

export const ACTIVATE_SUBSCRIPTION_EVENT = 'activate-subscription';
export const CLOSE_ACTIVATE_SUBSCRIPTION_BANNER_EVENT = 'close';

export default {
  name: 'SubscriptionActivationBanner',
  i18n: {
    bannerText: subscriptionBannerText,
    buttonText: activateCloudLicense,
    title: subscriptionBannerTitle,
  },
  components: {
    GlBanner,
    GlLink,
    GlSprintf,
  },
  inject: ['congratulationSvgPath', 'customersPortalUrl'],
  methods: {
    handleClose() {
      this.$emit(CLOSE_ACTIVATE_SUBSCRIPTION_BANNER_EVENT);
    },
    handlePrimary() {
      this.$emit(ACTIVATE_SUBSCRIPTION_EVENT);
    },
  },
};
</script>

<template>
  <gl-banner
    :button-text="$options.i18n.buttonText"
    :title="$options.i18n.title"
    variant="promotion"
    :svg-path="congratulationSvgPath"
    @close="handleClose"
    @primary="handlePrimary"
  >
    <p>
      <gl-sprintf :message="$options.i18n.bannerText">
        <template #blogPostLink="{ content }">
          <gl-link href="#" target="_blank">{{ content }}</gl-link>
        </template>
        <template #portalLink="{ content }">
          <gl-link :href="customersPortalUrl" target="_blank">{{ content }}</gl-link>
        </template>
      </gl-sprintf>
    </p>
  </gl-banner>
</template>
