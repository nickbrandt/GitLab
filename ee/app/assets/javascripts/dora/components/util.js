import { dataVizBlue500, dataVizOrange600 } from '@gitlab/ui/scss_to_js/scss_variables';
import dateFormat from 'dateformat';
import { merge, cloneDeep } from 'lodash';
import { getDatesInRange, nDaysBefore, getStartOfDay } from '~/lib/utils/datetime_utility';
import { s__ } from '~/locale';

/**
 * Converts the raw data fetched from the
 * [DORA Metrics API](https://docs.gitlab.com/ee/api/dora/metrics.html#get-project-level-dora-metrics)
 * into series data consumable by
 * [GlAreaChart](https://gitlab-org.gitlab.io/gitlab-ui/?path=/story/charts-area-chart--default)
 *
 * @param {Array} apiData The raw JSON data from the API request
 * @param {Date} startDate The first day (inclusive) of the graph's date range
 * @param {Date} endDate The last day (exclusive) of the graph's date range
 * @param {String} seriesName The name of the series
 * @param {*} emptyValue The value to substitute if the API data doesn't
 * include data for a particular date
 */
export const apiDataToChartSeries = (apiData, startDate, endDate, seriesName, emptyValue = 0) => {
  // Get a list of dates, one date per day in the graph's date range
  const beginningOfStartDate = getStartOfDay(startDate, { utc: true });
  const beginningOfEndDate = nDaysBefore(getStartOfDay(endDate, { utc: true }), 1, { utc: true });
  const dates = getDatesInRange(beginningOfStartDate, beginningOfEndDate).map((d) =>
    getStartOfDay(d, { utc: true }),
  );

  // Generate a map of API timestamps to its associated value.
  // The timestamps are explicitly set to the _beginning_ of the day (in UTC)
  // so that we can confidently compare dates by value below.
  const timestampToApiValue = apiData.reduce((acc, curr) => {
    const apiTimestamp = getStartOfDay(new Date(curr.date), { utc: true }).getTime();
    acc[apiTimestamp] = curr.value;
    return acc;
  }, {});

  // Fill in the API data (the API may exclude data points for dates that have no data)
  // and transform it for use in the graph
  const data = dates.map((date) => {
    const formattedDate = dateFormat(date, 'mmm d', true);
    return [formattedDate, timestampToApiValue[date.getTime()] ?? emptyValue];
  });

  return [
    {
      name: seriesName,
      data,
    },
  ];
};

/**
 * Linearly interpolates between two values
 *
 * @param {Number} valueAtT0 The value at t = 0
 * @param {Number} valueAtT1 The value at t = 1
 * @param {Number} t The current value of t
 *
 * @returns {Number} The result of the linear interpolation.
 */
const lerp = (valueAtT0, valueAtT1, t) => {
  return valueAtT0 * (1 - t) + valueAtT1 * t;
};

/**
 * Builds a second series that visually represents the "no data" (i.e. "null")
 * data points, and returns a new series Array that includes both the "null"
 * and "non-null" data sets.
 * This function returns new series data and does not modify the original instance.
 *
 * @param {Array} seriesData The lead time series data that has already been processed
 * by the `apiDataToChartSeries` function above.
 * @returns {Array} A new series Array
 */
export const buildNullSeriesForLeadTimeChart = (seriesData) => {
  const nonNullSeries = cloneDeep(seriesData[0]);

  // Loop through the series data and build a list of all the "gaps". A "gap" is
  // a section of the data set that only include `null` values. Each gap object
  // includes the start and end indices and the start and end values of the gap.
  const seriesGaps = [];
  let currentGap = null;
  nonNullSeries.data.forEach(([, value], index) => {
    if (value == null && currentGap == null) {
      currentGap = {};

      if (index > 0) {
        currentGap.startIndex = index - 1;
        const [, previousValue] = nonNullSeries.data[index - 1];
        currentGap.startValue = previousValue;
      }

      seriesGaps.push(currentGap);
    } else if (value != null && currentGap != null) {
      currentGap.endIndex = index;
      currentGap.endValue = value;
      currentGap = null;
    }
  });

  // Create a copy of the non-null series, but with all the data point values set to `null`
  const nullSeriesData = nonNullSeries.data.map(([date]) => [date, null]);

  // Render each of the gaps to the "null" series. Values are determined by linearly
  // interpolating between the start and end values.
  seriesGaps.forEach((gap) => {
    const startIndex = gap.startIndex ?? 0;
    const startValue = gap.startValue ?? gap.endValue ?? 0;
    const endIndex = gap.endIndex ?? nonNullSeries.data.length - 1;
    const endValue = gap.endValue ?? gap.startValue ?? 0;

    for (let i = startIndex; i <= endIndex; i += 1) {
      const t = (i - startIndex) / (endIndex - startIndex);
      nullSeriesData[i][1] = lerp(startValue, endValue, t);
    }
  });

  merge(nonNullSeries, {
    showSymbol: true,
    showAllSymbol: true,
    symbolSize: 8,
    lineStyle: {
      color: dataVizBlue500,
    },
    areaStyle: {
      color: dataVizBlue500,
    },
    itemStyle: {
      color: dataVizBlue500,
    },
  });

  const nullSeries = {
    name: s__('DORA4Metrics|No merge requests were deployed during this period'),
    data: nullSeriesData,
    lineStyle: {
      type: 'dashed',
      color: dataVizOrange600,
    },
    areaStyle: {
      color: 'none',
    },
    itemStyle: {
      color: dataVizOrange600,
    },
  };

  return [nullSeries, nonNullSeries];
};
