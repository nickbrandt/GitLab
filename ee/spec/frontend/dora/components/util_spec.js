import { apiDataToChartSeries, buildNullSeriesForLeadTimeChart } from 'ee/dora/components/util';

const lastWeekData = getJSONFixture(
  'api/dora/metrics/daily_lead_time_for_changes_for_last_week.json',
);

describe('ee/dora/components/util.js', () => {
  describe('apiDataToChartSeries', () => {
    it('transforms the data from the API into data the chart component can use', () => {
      const apiData = [
        // This is the date format we expect from the API
        { value: 5, date: '2015-06-28' },

        // But we should support _any_ date format
        { value: 1, date: '2015-06-28T20:00:00.000-0400' },
        { value: 8, date: '2015-07-01T00:00:00.000Z' },
      ];

      const startDate = new Date(2015, 5, 26, 10);
      const endDate = new Date(2015, 6, 4, 10);
      const chartTitle = 'Chart title';

      const expected = [
        {
          name: chartTitle,
          data: [
            ['Jun 26', 0],
            ['Jun 27', 0],
            ['Jun 28', 5],
            ['Jun 29', 1],
            ['Jun 30', 0],
            ['Jul 1', 8],
            ['Jul 2', 0],
            ['Jul 3', 0],
          ],
        },
      ];

      expect(apiDataToChartSeries(apiData, startDate, endDate, chartTitle)).toEqual(expected);
    });
  });

  describe('buildNullSeriesForLeadTimeChart', () => {
    it('returns series data with the expected styles and text', () => {
      const inputSeries = [
        {
          name: 'Chart title',
          data: [],
        },
      ];

      const expectedSeries = [
        {
          name: 'No merge requests were deployed during this period',
          data: expect.any(Array),
          lineStyle: {
            color: expect.any(String),
            type: 'dashed',
          },
          areaStyle: {
            color: 'none',
          },
          itemStyle: {
            color: expect.any(String),
          },
        },
        {
          name: 'Chart title',
          showAllSymbol: true,
          showSymbol: true,
          symbolSize: 8,
          data: expect.any(Array),
          lineStyle: {
            color: expect.any(String),
          },
          areaStyle: {
            color: expect.any(String),
          },
          itemStyle: {
            color: expect.any(String),
          },
        },
      ];

      expect(buildNullSeriesForLeadTimeChart(inputSeries)).toEqual(expectedSeries);
    });

    describe('series data', () => {
      describe('non-empty series', () => {
        it('returns the provided non-empty series data unmodified as the second series', () => {
          const inputSeries = [
            {
              data: [
                ['Mar 1', 4],
                ['Mar 2', null],
                ['Mar 3', null],
                ['Mar 4', 10],
              ],
            },
          ];

          const actualSeries = buildNullSeriesForLeadTimeChart(inputSeries);

          expect(actualSeries[1]).toMatchObject(inputSeries[0]);
        });
      });

      describe('empty series', () => {
        const compareSeriesData = (inputSeriesData, expectedEmptySeriesData) => {
          const actualEmptySeriesData = buildNullSeriesForLeadTimeChart([
            { data: inputSeriesData },
          ])[0].data;

          expect(actualEmptySeriesData).toEqual(expectedEmptySeriesData);
        };

        describe('when the data contains a gap in the middle of the data set', () => {
          it('builds the "no data" series by linealy interpolating between the provided data points', () => {
            const inputSeriesData = [
              ['Mar 1', 4],
              ['Mar 2', null],
              ['Mar 3', null],
              ['Mar 4', 10],
            ];

            const expectedEmptySeriesData = [
              ['Mar 1', 4],
              ['Mar 2', 6],
              ['Mar 3', 8],
              ['Mar 4', 10],
            ];

            compareSeriesData(inputSeriesData, expectedEmptySeriesData);
          });
        });

        describe('when the data contains a gap at the beginning of the data set', () => {
          it('fills in the gap using the first non-null data point value', () => {
            const inputSeriesData = [
              ['Mar 1', null],
              ['Mar 2', null],
              ['Mar 3', null],
              ['Mar 4', 10],
            ];

            const expectedEmptySeriesData = [
              ['Mar 1', 10],
              ['Mar 2', 10],
              ['Mar 3', 10],
              ['Mar 4', 10],
            ];

            compareSeriesData(inputSeriesData, expectedEmptySeriesData);
          });
        });

        describe('when the data contains a gap at the end of the data set', () => {
          it('fills in the gap using the last non-null data point value', () => {
            const inputSeriesData = [
              ['Mar 1', 10],
              ['Mar 2', null],
              ['Mar 3', null],
              ['Mar 4', null],
            ];

            const expectedEmptySeriesData = [
              ['Mar 1', 10],
              ['Mar 2', 10],
              ['Mar 3', 10],
              ['Mar 4', 10],
            ];

            compareSeriesData(inputSeriesData, expectedEmptySeriesData);
          });
        });

        describe('when the data contains all null values', () => {
          it('fills the empty series with all zeros', () => {
            const inputSeriesData = [
              ['Mar 1', null],
              ['Mar 2', null],
              ['Mar 3', null],
              ['Mar 4', null],
            ];

            const expectedEmptySeriesData = [
              ['Mar 1', 0],
              ['Mar 2', 0],
              ['Mar 3', 0],
              ['Mar 4', 0],
            ];

            compareSeriesData(inputSeriesData, expectedEmptySeriesData);
          });
        });
      });
    });
  });

  describe('lead time data', () => {
    it('returns the correct lead time chart data after all processing of the API response', () => {
      const chartData = buildNullSeriesForLeadTimeChart(
        apiDataToChartSeries(
          lastWeekData,
          new Date(2015, 5, 27, 10),
          new Date(2015, 6, 4, 10),
          'Lead time',
          null,
        ),
      );

      expect(chartData).toMatchSnapshot();
    });
  });
});
