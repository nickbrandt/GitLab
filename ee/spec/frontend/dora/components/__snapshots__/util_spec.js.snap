// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ee/dora/components/util.js lead time data returns the correct lead time chart data after all processing of the API response 1`] = `
Array [
  Object {
    "areaStyle": Object {
      "color": "none",
    },
    "data": Array [
      Array [
        "Jun 27",
        86400,
      ],
      Array [
        "Jun 28",
        86400,
      ],
      Array [
        "Jun 29",
        86400,
      ],
      Array [
        "Jun 30",
        86400,
      ],
      Array [
        "Jul 1",
        432000,
      ],
      Array [
        "Jul 2",
        518400,
      ],
      Array [
        "Jul 3",
        604800,
      ],
    ],
    "itemStyle": Object {
      "color": "#b24800",
    },
    "lineStyle": Object {
      "color": "#b24800",
      "type": "dashed",
    },
    "name": "No merge requests were deployed during this period",
  },
  Object {
    "areaStyle": Object {
      "color": "#5772ff",
    },
    "data": Array [
      Array [
        "Jun 27",
        null,
      ],
      Array [
        "Jun 28",
        null,
      ],
      Array [
        "Jun 29",
        null,
      ],
      Array [
        "Jun 30",
        86400,
      ],
      Array [
        "Jul 1",
        432000,
      ],
      Array [
        "Jul 2",
        null,
      ],
      Array [
        "Jul 3",
        604800,
      ],
    ],
    "itemStyle": Object {
      "color": "#5772ff",
    },
    "lineStyle": Object {
      "color": "#5772ff",
    },
    "name": "Lead time",
    "showAllSymbol": true,
    "showSymbol": true,
    "symbolSize": 8,
  },
]
`;
